File: apps/windows/SunoFlow/Tone.cs
using System;
using System.Collections.Generic;
using System.Drawing;

namespace SunoFlow;

/// <summary>
/// The writing voice a dictation is cleaned into — the Windows counterpart of
/// <c>Tone.swift</c>, and a deliberate 1:1 mirror of it.
///
/// The <see cref="Id"/> is what travels to the sidecar and on to the cleanup
/// gateway, which owns the closed set and the instruction behind each entry.
/// Nothing here describes <b>how</b> a voice writes — that lives server-side, so
/// a wording change ships as a gateway deploy rather than three app releases,
/// and the clients cannot drift from each other.
///
/// <see cref="Faithful"/> is the default and deliberately sends an empty id: the
/// gateway then builds the prompt it built before tones existed, and the user's
/// own wording survives untouched. Anyone who never presses the key is on
/// exactly the behaviour they have always had.
/// </summary>
internal sealed class Tone
{
    public string Id { get; }

    /// <summary>What the user sees — in the overlay, the tray menu, and Settings.</summary>
    public string Label { get; }

    /// <summary>One line of explanation, for the Settings row.</summary>
    public string Blurb { get; }

    /// <summary>
    /// One tint per voice, in the same muted family as the rest of the palette so
    /// the pill still reads as this product rather than as seven different ones.
    /// The values are the same hexes as the macOS build.
    ///
    /// Faithful is the app's own accent, unchanged: the default dictation looks
    /// exactly as it always has, and colour only becomes information once the
    /// user has chosen something. Colour is reinforcement, never the whole
    /// signal — seven hues are not learnable on their own, which is why a
    /// non-default voice also names itself next to the pill.
    /// </summary>
    public Color Tint { get; }

    private Tone(string id, string label, string blurb, Color tint)
    {
        Id = id;
        Label = label;
        Blurb = blurb;
        Tint = tint;
    }

    public static readonly Tone Faithful = new(
        "", "As spoken",
        "Your words, tidied. Filler and grammar only — nothing is rephrased.",
        Theme.Accent);

    public static readonly Tone Professional = new(
        "professional", "Professional",
        "Clear and polite, the way a colleague writes at work.",
        Color.FromArgb(42, 111, 176));   // #2A6FB0

    public static readonly Tone Formal = new(
        "formal", "Formal",
        "Serious and measured. No contractions, no slang.",
        Color.FromArgb(53, 86, 107));    // #35566B

    public static readonly Tone Casual = new(
        "casual", "Casual",
        "Relaxed and conversational, like writing to someone you know.",
        Color.FromArgb(192, 124, 44));   // #C07C2C

    public static readonly Tone Friendly = new(
        "friendly", "Friendly",
        "Warm and personable, without added enthusiasm.",
        Color.FromArgb(184, 82, 127));   // #B8527F

    public static readonly Tone Concise = new(
        "concise", "Concise",
        "As short as it can be without losing anything you said.",
        Color.FromArgb(30, 122, 85));    // #1E7A55

    public static readonly Tone Confident = new(
        "confident", "Confident",
        "Direct and decisive, with the hedging dropped.",
        Color.FromArgb(178, 74, 51));    // #B24A33

    /// <summary>
    /// Declaration order is the cycle order, and <see cref="Faithful"/> leads it
    /// so one press away from the default is always one press back to it.
    /// </summary>
    public static readonly IReadOnlyList<Tone> All = new[]
    {
        Faithful, Professional, Formal, Casual, Friendly, Concise, Confident,
    };

    /// <summary>The next voice in the cycle.</summary>
    public Tone Next()
    {
        int i = 0;
        for (int n = 0; n < All.Count; n++)
            if (ReferenceEquals(All[n], this)) { i = n; break; }
        return All[(i + 1) % All.Count];
    }

    /// <summary>
    /// Read a stored id, falling back to the default. A value this build does not
    /// know — a settings file written by a newer version, or edited by hand —
    /// leaves the user's words alone rather than guessing at a voice.
    /// </summary>
    public static Tone From(string? id)
    {
        if (!string.IsNullOrEmpty(id))
            foreach (var t in All)
                if (string.Equals(t.Id, id, StringComparison.OrdinalIgnoreCase))
                    return t;
        return Faithful;
    }

    public override string ToString() => Label;
}

File: SunoFlowApp/Sources/SunoFlow/Tone.swift
import Cocoa
import SwiftUI

/// The writing voice a dictation is cleaned into.
///
/// The `rawValue` is the ID sent to the sidecar and on to the cleanup gateway,
/// which owns the closed set and the instruction behind each entry. Nothing
/// here describes *how* a voice writes — that lives server-side, so a wording
/// change ships as a gateway deploy rather than an app release, and the three
/// clients cannot drift from each other.
///
/// `faithful` is the default and deliberately sends nothing: an empty ID means
/// the gateway builds the prompt it built before tones existed, and the user's
/// own wording survives untouched. Anyone who never presses the key is on
/// exactly the behaviour they have always had.
enum Tone: String, CaseIterable, Identifiable {
    case faithful = ""
    case professional
    case formal
    case casual
    case friendly
    case concise
    case confident

    var id: String { rawValue }

    /// What the user sees — in the overlay, the menu, and Settings.
    var label: String {
        switch self {
        case .faithful:     return "As spoken"
        case .professional: return "Professional"
        case .formal:       return "Formal"
        case .casual:       return "Casual"
        case .friendly:     return "Friendly"
        case .concise:      return "Concise"
        case .confident:    return "Confident"
        }
    }

    /// One line of explanation, for the Settings list.
    var blurb: String {
        switch self {
        case .faithful:     return "Your words, tidied. Filler and grammar only — nothing is rephrased."
        case .professional: return "Clear and polite, the way a colleague writes at work."
        case .formal:       return "Serious and measured. No contractions, no slang."
        case .casual:       return "Relaxed and conversational, like writing to someone you know."
        case .friendly:     return "Warm and personable, without added enthusiasm."
        case .concise:      return "As short as it can be without losing anything you said."
        case .confident:    return "Direct and decisive, with the hedging dropped."
        }
    }

    /// The next voice in the cycle. Declaration order is the cycle order, and
    /// `faithful` leads it so one press away from the default is always one
    /// press back to it.
    var next: Tone {
        let all = Tone.allCases
        let i = all.firstIndex(of: self) ?? 0
        return all[(i + 1) % all.count]
    }

    /// Read a stored or received ID, falling back to the default. A value the
    /// app does not know — a preference written by a newer build, or edited by
    /// hand — leaves the user's words alone rather than guessing at a voice.
    static func from(_ raw: String?) -> Tone {
        Tone(rawValue: raw ?? "") ?? .faithful
    }
}

// MARK: - Colour

/// One tint per voice, in the same muted family as the rest of the palette so
/// the pill still reads as this product rather than as seven different ones.
///
/// `faithful` is the app's own accent, unchanged: the default dictation looks
/// exactly as it always has, and colour only becomes information once the user
/// has chosen something.
///
/// Colour is reinforcement, never the whole signal — seven hues are not
/// learnable on their own, which is why a non-default voice also names itself
/// in a chip below the pill for as long as it is up. See `BubbleView.setTone`.
extension Tone {
    var tint: Color {
        switch self {
        case .faithful:     return Theme.accent
        case .professional: return Color(red: 0.165, green: 0.435, blue: 0.690)  // #2A6FB0
        case .formal:       return Color(red: 0.208, green: 0.337, blue: 0.420)  // #35566B
        case .casual:       return Color(red: 0.753, green: 0.486, blue: 0.173)  // #C07C2C
        case .friendly:     return Color(red: 0.722, green: 0.322, blue: 0.498)  // #B8527F
        case .concise:      return Color(red: 0.118, green: 0.478, blue: 0.333)  // #1E7A55
        case .confident:    return Color(red: 0.698, green: 0.290, blue: 0.200)  // #B24A33
        }
    }

    var nsTint: NSColor { NSColor(tint) }
}
